### DISCUSSION TITLE
`p-2` overrides `p-6` but `p-7` overrides `p-2`?

### DISCUSSION THREAD

**Discussion by rafeautie** [tailwindlabs/tailwindcss#6730] [Help] [answered]
Hi,

I need help understanding what is happening. Title says it all. Screenshots to illustrate the issue.

**Problem statement**: `p-2` overrides `p-6` but `p-7` overrides `p-2`.

I do not understand why this happens.

**Screenshots:**
![Screen Shot 2021-12-25 at 3 47 53 AM](https://user-images.githubusercontent.com/45886348/147384257-f476ddf8-3d84-4d1d-abf2-f859ce2e542a.png)
Above screenshot is of `p-2` overriding `p-6`.

![Screen Shot 2021-12-25 at 3 49 44 AM](https://user-images.githubusercontent.com/45886348/147384292-2e2d6f3c-53ef-4b08-bd85-c92fd7ed1572.png)
Above screenshot is of `p-7` overriding `p-2`.

**Note/Things Ive Tried:**
- Order of styles applied does not affect anything.
- In case of first screenshot, toggling `padding: 0.5rem` enables `padding: 1.5rem`. So Im 99% my understanding of the issue is correct. Please let me know if Im misunderstanding something.

To add, my use case is creating a core component library and needing to override default styles. `p-2` being the default style.

### Update:

Ok this seems like a bug but Im not sure. Every time i restart the bundler, the style that gets overridden changes. After restarting, now p-2 overrides p-7.

**Comment by adamwathan** [comment, marked as answer]
Hey! So what you are finding is that the order of classes in the same "category" are non-deterministic which is by design for performance reasons.

Classes like `p-2` and `p-7` are considered in conflict with each other when both present, much like `text-left` and `text-right` for example.

You'll see on Tailwind Play and in the Tailwind CSS IntelliSense plugin we show a warning for these situations:

https://play.tailwindcss.com/emjlpJLBL1

![image](https://user-images.githubusercontent.com/4323180/147389202-6155b0a6-9981-4d64-8908-10fd945b14d8.png)

So the solution here is just don't add conflicting classes, as the behavior is considered undefined. What's actually happening is Tailwind is generating the classes based on which ones it sees first in your templates, so it all depends on what order the template files are processed in.

Hope that helps! 🎄

**Reply by rafeautie** [reply]
Ah yes that makes sense. I didn't see the warning in vscode because the className props are being combined programatically. Thanks for this. I suppose this is a flaw with how i'm creating the core component library for my project. This makes me wonder if there are recommended patterns for component libraries built with tailwind. How do people handle the scenario where, for example, i want to override a buttons default padding? Would I just need to be conscious of the styles i'm using or is there a proper solution that doesn't required the extra mental effort? :)

**Reply by simonmaass** [reply]
I just came accross the same problem... I was expecting an element having py-3 for example and I add a py-5... that the "bigger" style will override the smaller one...
